core/domain/prod_validation_jobs_one_off.py

```diff
@@ -529,6 +544,117 @@ def validate(cls, item):
         cls._validate_post_commit_is_private(item)
 
 
+class BaseUserModelValidator(BaseModelValidator):
+    """Class for validating BaseUserModels."""
+
+    @classmethod
+    def _get_model_id_regex(cls, unused_item):
+        return '^\\d*$'
+
+    @classmethod
+    def _get_exp_ids(cls, unused_item):
+        """Returns a list of exploration ids related to the user model.
+
+        Args:
+            unused_item: ndb.Model. BaseUserModel to validate.
+
+        Returns:
+            list(str). List of exploration ids related to the model.
+        """
+        return []
+
+    @classmethod
+    def _get_col_ids(cls, unused_item):
+        """Returns a list of collection ids related to the user model.
+
+        Args:
+            unused_item: ndb.Model. BaseUserModel to validate.
+
+        Returns:
+            list(str). List of collection ids related to the model.
+        """
+        return []
+
+    @classmethod
+    def _validate_explorations_are_public(cls, item):
+        """Validates that explorations for model are public.
+
+        Args:
+            item: ndb.Model. BaseUserModel to validate.
+        """
+        exp_ids = cls._get_exp_ids(item)
+        private_exp_ids = [
+            exp_id for exp_id in exp_ids if (
+                rights_manager.is_exploration_private(exp_id))]
+        if private_exp_ids:
+            cls.errors['public exploration check'].append(
+                'Entity id %s: Explorations with ids %s are private' % (
+                    item.id, private_exp_ids))
+
+    @classmethod
+    def _validate_collections_are_public(cls, item):
+        """Validates that collections for model are public.
+
+        Args:
+            item: ndb.Model. BaseUserModel to validate.
+        """
+        col_ids = cls._get_col_ids(item)
+        private_col_ids = [
+            col_id for col_id in col_ids if (
+                rights_manager.is_collection_private(col_id))]
+        if private_col_ids:
+            cls.errors['public collection check'].append(
+                'Entity id %s: Collections with ids %s are private' % (
+                    item.id, private_col_ids))
+
+    @classmethod
+    def _get_common_properties_of_external_model_which_should_not_match(
+            cls, unused_item):
+        """Returns a list common properties to dismatch.
```

Comment:
I don't really understand this docstring. Can you elaborate on what "dismatch" / "should not match" means? What are the things that shouldn't match, exactly?

Reply:
Updated, added an example to explain.

If two models store opposite features, the common properties in both the models should not have the same values. For example, in `CompletedActivitiesModel` and `IncompleteActivitiesModel` the `exploration_ids` in one model should not match those in the other model.

Reply:
I am using `get_by_id` at two other places:

1. https://github.com/oppia/oppia/pull/7023/files#diff-ecafbbaeb117bcae187e78e5b33e120fR3989

This to obtain exploration models corresponding to node ids in StoryProgressModel. For this, I wil need to fetch story models and the get the exploration ids for completed nodes in StoryProgressModel. So, I don't think if we can separate this.

2. https://github.com/oppia/oppia/pull/7023/files#diff-ecafbbaeb117bcae187e78e5b33e120fR4080

This to obtain UserBulkEmailModels corresponding to recipients to whom bulk email is sent after a they qualify a query. Here also, I need to fetch the BulkEmailModel for each entity and iterate over recipients to fetch user bulk email models. So, the case becomes same as above.

What should I do in these two cases?

Reply:
I am not entirely sure how your comment relates to my comment, but one observation I have is that you might want to use get_multi() in both cases rather than lots of get_by_id()'s in a for loop. The latter will take much longer.

Reply:
Oh, sorry I added this in the wrong thread. I wanted to comment this [here](https://github.com/oppia/oppia/pull/7023#discussion_r297532713). I will change `get_by_id` to `get_multi`.

Reply:
Just a note -- it's the repeated query() that we should try to avoid if possible (especially doing N queries when 1 will do).

get_by_id() is fast, but if you're doing multiple of them, do a single get_multi() instead.

The above principles should be sufficient for most cases.

Reply:
Updated to `get_multi`.